microsoft/BotFramework-WebChat
# Fix usability on accessibility

> Fix #3075.

## Changelog Entry

### Fixed

-  Fixes [#3075](https://github.com/microsoft/BotFramework-WebChat/issues/3075). Fix usability issues around accessibility, by [@compulim](https://github.com/compulim) in PR [#3076](https://github.com/microsoft/BotFramework-WebChat/issue/3076)
   -  Fix timestamp should not be narrated more than once.
   -  Associate the activity text with its attachments, by adding a `role="region"` to the activity DOM element.

## Description

(See changelog above)

This may be a temporary fix until we found a better DOM structure.

## Specific Changes

- Update `<StackLayout>` to narrate as `aria-label` on the activity root DOM element
- Update `<CarouselFilmstrip>` to match `<StackLayout>` on screen reader text
- Update `aria-relevant` for `<BasicTranscript>` to only narrate additions of node
- Remove `<Timestamp>` from narration, because it will be narrated by the activity

## Test

1. Run Windows Narrator (<kbd>Windows</kbd> + <kbd>CTRL</kbd> + <kbd>ENTER</kbd>)
1. Say "hello" to the bot
   - Expect: The narration should be minimal and not repetitive
1. Say "card inputs breakfast"
1. Use mouse to focus on any box of "inputs card"
1. Use mouse to focus on the button on "breakfast card"
   - Expect: The narration should mention it's on breakfast card
1. Use mouse to focus on the dropbox of "breakfast card"
   - Expect: The narration should *not* mention it's on breakfast card again
1. Use mouse to focus on any input on "card inputs"
   - Expect: The narration should mention it's on inputs card

---

-  [x] ~Testing Added~
   -  No tests added because accessibility cannot be automated
